WireDomain/Sources/WireDomain/Repositories/Message/MessageLocalStore.swift

```diff
@@ -363,24 +490,625 @@ public final class MessageLocalStore: MessageLocalStoreProtocol {
+    ) async {
+        let selfUser = await userLocalStore.fetchSelfUser()
+
+        let isSelf = await context.perform {
+            conversation.isSelfConversation && senderID != selfUser.remoteIdentifier
+        }
+
+        let conversationID = await context.perform {
+            conversation.remoteIdentifier
+        }
+
+        var logAttributes: LogAttributes = [
+            .messageType: messageType,
+            .conversationId: conversationID?.safeForLoggingDescription
+        ]
+
+        guard !isSelf else {
+            return WireLogger.eventProcessing.debug(
+                "Illegal sender or conversation, abort processing.",
+                attributes: logAttributes
+            )
+        }
+
+        let isForcedReadOnly = await conversationLocalStore.isConversationForcedReadOnly(
+            conversation
+        )
+
+        guard !isForcedReadOnly else {
+            return WireLogger.eventProcessing.warn(
+                "Ignoring incoming message in readonly conversation.",
+                attributes: logAttributes
+            )
+        }
+
+        guard let (genericMessage, content) = await getGenericMessage(
+            from: message,
+            externalData: externalData,
+            senderID: senderID,
+            senderDomain: senderDomain,
+            date: date,
+            conversation: conversation,
+            logAttributes: logAttributes
+        ) else {
+            return
+        }
+
+        WireLogger.eventProcessing.debug("Processing:\n\(genericMessage)")
+        logAttributes[.nonce] = UUID(uuidString: genericMessage.messageID) ?? "<nil>"
+        WireLogger.eventProcessing.debug("Processing message", attributes: logAttributes)
+
+        // Update the legal hold state in the conversation
+        await context.perform {
+            conversation.updateSecurityLevelIfNeededAfterReceiving(
+                message: genericMessage,
+                timestamp: date ?? .now
+            )
+        }
+
+        // Add sender to the conversation if needed
+        await addParticipantIfNeeded(
+            senderID: senderID,
+            senderDomain: senderDomain,
+            in: conversation,
+            date: date
+        )
+
+        // Process the message and its content
+        await processMessageContent(
+            genericMessage,
+            content: content,
+            in: conversation,
+            senderID: senderID,
+            senderDomain: senderDomain,
+            senderClientID: senderClientID,
+            logAttributes: logAttributes,
+            date: date
+        )
+    }
+
+    private func processMessageContent(
+        _ message: GenericMessage,
+        content: GenericMessage.OneOf_Content,
+        in conversation: ZMConversation,
+        senderID: UUID,
+        senderDomain: String,
+        senderClientID: String?,
+        logAttributes: LogAttributes,
+        date: Date?
+    ) async {
+        await context.perform { [self] in
```

Comment:
I wouldn't expect to see processing inside the store, rather only actual message insertions. Some of these message types are processed and discarded, which I feel like should be done in the event processor level. What do you think?

Reply:
done in https://github.com/wireapp/wire-ios/commit/c08ed17ebb783810d1b65e98d08654d10a6aaf59, revisited the current implementation